Forms: add "empty trash" button

simison opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Resolves https://github.com/Automattic/jetpack/issues/41539
Resolves FORMS-212

Old classic view used to have "Empty Trash" provided by core UI — dataviews doesn't have this ability built in, so we need to build it custom:

![image](https://github.com/user-attachments/assets/9a1b782d-8dc9-45bd-83c4-1d4e01991816)


## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Add "empty trash" button

<img width="1423" alt="Screenshot 2025-07-08 at 13 45 27" src="https://github.com/user-attachments/assets/32ea9c06-717b-4fc8-a551-15d32d832f12" />

When trash is already empty, button is disabled with tooltip on hover:

<img width="1421" alt="Screenshot 2025-07-08 at 13 44 55" src="https://github.com/user-attachments/assets/6724d9c3-6bc3-49f1-8292-da63a63260bb" />

When clicking, it asks for confirmation:

<img width="1357" height="932" alt="Screenshot 2025-07-11 at 15 04 29" src="https://github.com/user-attachments/assets/01dd5cac-6bdb-4075-842b-88cae75e15e3" />

Whole flow:


https://github.com/user-attachments/assets/c26341a2-cf8c-4e49-ba29-a0b4c59b450d




### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

simison committed:
Forms: add empty trash

simison committed:
Changelogs

simison committed:
Use WP_Query

simison commented on projects/packages/forms/src/dashboard/components/empty-trash-button/index.jsx:
```diff
@@ -0,0 +1,89 @@
+			path: `/wp/v2/feedback/trash`,
+		} )
+			.then( response => {
+				invalidateFilters(); // Reload the view
```
This should reset the view if I understood correct but it doesn't. 🤔

simison commented on projects/packages/forms/src/dashboard/components/empty-trash-button/index.jsx:
```diff
@@ -0,0 +1,89 @@
+		status: 'trash',
+	} );
+
+	const isEmpty = totalItems === 0;
```
It would be useful to query for count elsewhere so that we could re-use counts from `components/inbox-status-toggle`, and apply "disabled" state for Export buttons as well.

Seemed like a fine thing to do in a separate PR either before or after this PR.

Update: done in https://github.com/Automattic/jetpack/pull/44293

CGastrell committed:
use invalidateResolutionForStore after emptying trash

simison committed:
Empty trash button on mobile

simison committed:
Typescriptify

CGastrell committed:
Forms: add empty trash test (#44238)

* add test file for empty trash button

* update lock file

* use the proper js test script

* delete unused test script

* update composer lock file

* convert file to typescript and fix type errors

* add jetpack changelog

---------

Co-authored-by: Douglas <douglas.henri@automattic.com>

CGastrell commented:
So eager to get this rolling, I never realized we don't confirm with the user in any way and this is a quite destructive action, even for trash. Maybe get a confirmation modal before actually performing the request?

CGastrell committed:
add modal confirmation before triggering the request to empty trash

simison commented:
> So eager to get this rolling, I never realized we don't confirm with the user in any way and this is a quite destructive action, even for trash. Maybe get a confirmation modal before actually performing the request?

Seems fine; tested Gmail/Drive and they use confirm too, although paired with automatically emptying trash every 30 days.

I had originally thought to leave confirmation out, since trash is 2nd place after original deletion, and I generally prefer "undo" actions over confirmations, but that's harder to do here.

simison commented on projects/packages/forms/src/dashboard/components/empty-trash-button/style.scss:
```diff
@@ -0,0 +1,5 @@
+.jp-forms__empty-trash-modal-actions {
+	display: flex;
+	gap: 8px;
+	justify-content: flex-end;
```
This kind of CSS is usually possible easier with HStack/Vstack ;-)

simison commented on projects/packages/forms/package.json:
```diff
@@ -82,6 +81,7 @@
 		"@babel/core": "7.27.7",
 		"@babel/runtime": "7.27.6",
 		"@testing-library/dom": "10.4.0",
+		"@testing-library/react": "^14.2.1",
```
I'll update this version to match recent PR:
- https://github.com/Automattic/jetpack/pull/44217

simison committed:
Update @testing-library/react

simison committed:
Update @testing-library/user-event

simison committed:
Use and update userEvent

simison committed:
update @testing-library/user-event

simison committed:
Disable failing test

simison commented on projects/packages/forms/tests/js/dashboard/components/empty-trash-button/index.test.jsx:
```diff
@@ -0,0 +1,127 @@
+	} );
+
+	// eslint-disable-next-line jest/no-disabled-tests
+	it.skip( 'shows loading state while emptying trash', async () => {
```
@CGastrell this test kept failing for me so I disabled it for now:

```
  ● EmptyTrashButton › shows loading state while emptying trash

    Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.

    Check the render method of `EmptyTrashButton`.

      112 |
      113 | 		const button = screen.getByText( 'Empty trash' );
    > 114 | 		await userEvent.click( button );
```

Do you know what might be up?

The test was failing already before but I switched to use `userEvent` instead of deprecated `fireEvent` to see if it helps (it didn't) but feels like better not add deprecated methods to new code.

CGastrell replied:
Issue was with how mocking the dispatchers was done, I asked Claude to fix it

simison committed:
Switch to ConfirmDialog and update copy

simison commented:
@CGastrell thanks for adding confirm modal! I simplified it quite a bit by using core's "ConfirmModal" instead of Modal+buttons — we lose the "danger" button which would be nice to keep, but that seems minor.

I also updated the copy to not be so repetative and focus more on _outcome_ (deleting) than _action_ (confirming).

Before

<img width="758" height="317" alt="Screenshot 2025-07-11 at 12 47 38" src="https://github.com/user-attachments/assets/0217f8fa-8448-45c7-832f-141569b2dd8e" />

After

<img width="591" height="300" alt="Screenshot 2025-07-11 at 12 51 32" src="https://github.com/user-attachments/assets/e990b117-13bd-4eb7-940a-62ba9ddd2d51" />

ilonagl commented:
Great work, @simison! 

Thoughts: 
- A confirmation modal should also appear when users permanently delete items using the multi-select delete action from the bottom left. Right now, it happens automatically without confirmation.
<img width="1512" height="982" alt="Screenshot 2025-07-11 at 13 30 26" src="https://github.com/user-attachments/assets/85c91e22-ad06-4add-bdf3-24529eea3245" />
- Could we identify how many items are in trash by showing it at the top button (e.g., 'Empty trash (1)')? Same as we do for Export. It helps to understand what will be affected by this action. 
<img width="1512" height="982" alt="Screenshot 2025-07-11 at 13 25 44" src="https://github.com/user-attachments/assets/46b56a8c-4fbe-47e2-8669-999382e97c1d" />

 
Copy: 
- We can't say that all responses will be deleted, because maybe I chose only one of the items to be permanently deleted only? Example below: 
<img width="1512" height="982" alt="Screenshot 2025-07-11 at 13 38 29" src="https://github.com/user-attachments/assets/7faffdcb-558f-4fb0-9407-156b65edf66c" />

Here are some examples of similar modals: 
<img width="1920" height="1320" alt="image" src="https://github.com/user-attachments/assets/a1b0dc8d-76ad-413d-8823-984c5d29a4e3" />
<img width="1920" height="1320" alt="image" src="https://github.com/user-attachments/assets/90e32158-3676-4d4b-8a87-35b951986789" />
<img width="1920" height="1320" alt="image" src="https://github.com/user-attachments/assets/68a5cea5-3f22-435f-afb4-b0c269efb78e" />
<img width="1920" height="1320" alt="image" src="https://github.com/user-attachments/assets/3a86fae6-8e2c-4d60-8bda-7c5c8ba8ace1" />
<img width="1920" height="1320" alt="image" src="https://github.com/user-attachments/assets/352fe913-4dfe-4719-85b1-82d108e2a8f5" />
<img width="1920" height="1320" alt="image" src="https://github.com/user-attachments/assets/ad5cd5bd-e2aa-4682-8829-50e2adb56306" />
<img width="1920" height="1320" alt="image" src="https://github.com/user-attachments/assets/481fe3fc-7606-4927-90a8-a90d397f8ea7" />

Copy suggestions: 

Header: Delete forever
Text: All responses in trash will be deleted forever. This action cannot be undone. 
Actions: Cancel | Delete

simison committed:
Update copy

jsnajdr commented on projects/packages/forms/src/dashboard/components/empty-trash-button/index.tsx:
```diff
@@ -0,0 +1,128 @@
+				setIsLoading( false );
+			} )
+			.finally( () => {
+				invalidateResolutionForStore( dashboardStore );
```
Here you would rather do a targeted invalidation:
```js
invalidateResolution( 'getEntityRecords', [ 'postType', 'feedback', { status: 'trash' } ] )
```
but it doesn't work. Why? After emptying the trash and invalidating the entire `coreStore`, I see the following requests:
```
/wp-json/wp/v2/feedback?context=edit&search=&page=1&per_page=1&status=publish%2Cdraft&_fields=id
/wp-json/wp/v2/feedback?context=edit&search=&page=1&per_page=1&status=spam&_fields=id
/wp-json/wp/v2/feedback?context=edit&search=&page=1&per_page=1&status=trash&_fields=id
/wp-json/wp/v2/feedback?context=edit&per_page=20&page=1&search=&status=trash
/wp-json/wp/v2/feedback?context=edit&status=trash
```
Three of them fetch the `trash` items. It's basically the same info, but it's fetched three times.

The first three requests are from the `InboxStatusToggle` component, where the three queries have custom params like `per_page=1`, because we want only the total count, so it makes the response body minimal.

The 4th request, with `per_page=20` is done by the `InboxView` component when fetching data for the dataview.

The last request, with minimal params, is issued by the `EmptyTrashButton` component.

If in the `EmptyTrashButton` you call
```js
invalidateResolution( 'getEntityRecords', [ 'postType', 'feedback', { status: 'trash' } ] )
```
it invalidates only the single query that has exactly the `{ status: 'trash' }` params. The other two queries have different params, so they are not affected. That's why neither the dataview nor the inbox status will get updated. And it also explains why the full invalidation of the entire store solves the problem.

The best solution is to consolidate the three requests into one. The `totalCount` needed by `InboxStatusToggle` is exactly the same as the `totalCount` needed by `EmptyTrashButton`.

And if the `InboxView` always loads the full data for the dataview, the other two components that need only `totalCount` should somehow reuse the `trash` query.

simison replied:
I noted [earlier](https://github.com/Automattic/jetpack/pull/44225#discussion_r2192181303) that we'd ideally fetch the count just once in more central hook but in my mind we can do that in a follow-up after this PR, or before this PR but in separate PR regardless.

simison replied:
Consolidated in:
- https://github.com/Automattic/jetpack/pull/44293

jsnajdr commented on projects/packages/forms/src/dashboard/components/empty-trash-button/index.tsx:
```diff
@@ -0,0 +1,128 @@
+		status: 'trash',
+	} );
+
+	const isEmpty = totalItems === 0;
```
This causes a little bug: when loading the `jetpack-forms-admin` page, the "Empty trash" button is always enabled for a moment while the `/feedback` data are loading, and gets disabled only when the response arrives. That's because `totalItems === undefined` during the load. The `isEmpty` condition should also cover this case and enable the button only when `totalItems` is really `> 0`.

simison replied:
Thanks! I addressed this in https://github.com/Automattic/jetpack/pull/44225/commits/7e2dc9fd9a96071de4cdff78465f8233e01ad950

I left the button enabled even when we don't know if the trash is truly empty, because there's no real harm letting user click the button in empty bin, and there's less jumpiness between enabled/disabled as the counter loads.

simison committed:
Take into account selected items in modal copy

simison committed:
Use useDispatch

simison committed:
Use more logical names for confirm dialog state

simison committed:
Hide confirm dialog when starting to empty

CGastrell committed:
fix tests
